File: paperboy/src/main/java/com/github/porokoro/paperboy/PaperboyAdapter.java
/*
 * Copyright (C) 2015 porokoro
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package com.github.porokoro.paperboy;

import android.content.Context;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.support.annotation.ColorInt;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.Html;
import android.text.Spanned;
import android.text.method.LinkMovementMethod;
import android.text.style.ClickableSpan;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import static android.support.v7.widget.RecyclerView.Adapter;
import static android.support.v7.widget.RecyclerView.ViewHolder;


class PaperboyAdapter extends Adapter<ViewHolder> {
    private static final int    DEFAULT_COLOR = Color.BLACK;
    private static final String TAG           = PaperboyAdapter.class.getSimpleName();

    private final LayoutInflater        m_inflater;
    private final List<Element>         m_dataset;
    @ElementType
    private final int                   m_elementType;
    private final PaperboyConfiguration m_config;

    public PaperboyAdapter(@NonNull Context context, @NonNull PaperboyConfiguration config) {
        m_inflater = LayoutInflater.from(context);
        m_dataset = new ArrayList<>();
        m_config = config;
        m_elementType = getElementType(config.getViewType());
    }

    public void setData(@NonNull List<PaperboySection> dataset) {
        for (PaperboySection section : dataset) {
            if (m_config.isSortItems() || m_config.getViewType() == ViewTypes.HEADER) {
                Collections.sort(section.getItems(), new Comparator<PaperboyItem>() {
                    @Override
                    public int compare(@NonNull PaperboyItem lhs, @NonNull PaperboyItem rhs) {
                        ItemType lhsDef = m_config.getItemTypes().get(lhs.getType());
                        ItemType rhsDef = m_config.getItemTypes().get(rhs.getType());
                        return (lhsDef != null ? lhsDef.getSortOrder() : Integer.MAX_VALUE) -
                                (rhsDef != null ? rhsDef.getSortOrder() : Integer.MAX_VALUE);
                    }
                });
            }

            Element sectionElement = new Element();
            sectionElement.type = ElementTypes.SECTION_HEADER;
            sectionElement.data = section;
            m_dataset.add(sectionElement);

            int prevItemType = DefaultItemTypes.NONE;

            for (PaperboyItem item : section.getItems()) {
                if (item.getType() != prevItemType && m_config.getViewType() == ViewTypes.HEADER) {
                    Element typeElement = new Element();
                    typeElement.type = ElementTypes.TYPE_HEADER;
                    typeElement.data = m_config.getItemTypes().get(item.getType());
                    m_dataset.add(typeElement);

                    prevItemType = item.getType();
                }

                Element itemElement = new Element();
                itemElement.type = m_elementType;
                itemElement.data = item;
                m_dataset.add(itemElement);
            }
        }

        notifyDataSetChanged();
    }

    private static class ViewHolderSection extends ViewHolder {
        TextView name;

        public ViewHolderSection(@NonNull View itemView) {
            super(itemView);
            name = (TextView) itemView.findViewById(R.id.name);
        }
    }

    private static class ViewHolderType extends ViewHolder {
        TextView name;

        public ViewHolderType(@NonNull View itemView) {
            super(itemView);
            name = (TextView) itemView.findViewById(R.id.name);
        }
    }

    private static class ViewHolderItemNone extends ViewHolder {
        TextView title;

        public ViewHolderItemNone(@NonNull View itemView) {
            super(itemView);
            title = (TextView) itemView.findViewById(R.id.title);
        }
    }

    private static class ViewHolderItemLabel extends ViewHolder {
        TextView type;
        TextView title;

        public ViewHolderItemLabel(@NonNull View itemView) {
            super(itemView);
            type = (TextView) itemView.findViewById(R.id.type);
            title = (TextView) itemView.findViewById(R.id.title);
        }
    }

    private static class ViewHolderItemIcon extends ViewHolder {
        ImageView type;
        TextView  title;

        public ViewHolderItemIcon(@NonNull View itemView) {
            super(itemView);
            type = (ImageView) itemView.findViewById(R.id.type);
            title = (TextView) itemView.findViewById(R.id.title);
        }
    }

    private static class Element {
        @ElementType
        int type;
        Object data;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, @ElementType int viewType) {
        switch (viewType) {
            case ElementTypes.SECTION_HEADER: {
                int sectionLayout = R.layout.list_item_paperboy_section;
                if (m_config.getSectionLayout() != 0)
                    sectionLayout = m_config.getSectionLayout();

                View rootView = m_inflater.inflate(sectionLayout, parent, false);
                return new ViewHolderSection(rootView);
            }
            case ElementTypes.TYPE_HEADER: {
                int typeLayout = R.layout.list_item_paperboy_type;
                if (m_config.getTypeLayout() != 0)
                    typeLayout = m_config.getTypeLayout();

                View rootView = m_inflater.inflate(typeLayout, parent, false);
                return new ViewHolderType(rootView);
            }
            case ElementTypes.ITEM_NONE: {
                View rootView = inflateItemView(R.layout.list_item_paperboy_item_none, parent);
                return new ViewHolderItemNone(rootView);
            }
            case ElementTypes.ITEM_LABEL: {
                View rootView = inflateItemView(R.layout.list_item_paperboy_item_label, parent);
                return new ViewHolderItemLabel(rootView);
            }
            case ElementTypes.ITEM_ICON: {
                View rootView = inflateItemView(R.layout.list_item_paperboy_item_icon, parent);
                return new ViewHolderItemIcon(rootView);
            }
            default:
                return null;
        }
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        if (holder instanceof ViewHolderSection) {
            ViewHolderSection viewHolder = (ViewHolderSection) holder;
            PaperboySection data = (PaperboySection) m_dataset.get(position).data;

            if (viewHolder.name == null)
                Log.w(TAG, "View id 'R.id.name' missing in custom layout");
            else
                viewHolder.name.setText(data.getName());
        }
        else if (holder instanceof ViewHolderType) {
            ViewHolderType viewHolder = (ViewHolderType) holder;
            ItemType data = (ItemType) m_dataset.get(position).data;

            if (viewHolder.name == null)
                Log.w(TAG, "View id 'R.id.name' missing in custom layout");
            else {
                if (data == null)
                    viewHolder.name.setVisibility(View.GONE);
                else {
                    viewHolder.name.setVisibility(View.VISIBLE);
                    viewHolder.name.setText(data.getTitlePlural());
                    viewHolder.name.setTextColor(data.getColor());
                }
            }
        }
        else if (holder instanceof ViewHolderItemNone) {
            ViewHolderItemNone viewHolder = (ViewHolderItemNone) holder;
            PaperboyItem data = (PaperboyItem) m_dataset.get(position).data;

            if (viewHolder.title == null)
                Log.w(TAG, "View id 'R.id.title' missing in custom layout");
            else
                setHtmlText(viewHolder.title, data.getTitle());
        }
        else if (holder instanceof ViewHolderItemLabel) {
            ViewHolderItemLabel viewHolder = (ViewHolderItemLabel) holder;
            PaperboyItem data = (PaperboyItem) m_dataset.get(position).data;
            ItemType definition = m_config.getItemTypes().get(data.getType());

            if (viewHolder.type == null)
                Log.w(TAG, "View id 'R.id.type' missing in custom layout");
            else {
                if (definition == null)
                    viewHolder.type.setVisibility(View.GONE);
                else {
                    viewHolder.type.setVisibility(View.VISIBLE);
                    viewHolder.type.setText(definition.getTitleSingular());
                    viewHolder.type.setBackgroundResource(R.drawable.paperboy_label_background);
                    viewHolder.type.getBackground().setColorFilter(getColor(definition), PorterDuff.Mode.SRC_ATOP);
                }
            }

            if (viewHolder.title == null)
                Log.w(TAG, "View id 'R.id.title' missing in custom layout");
            else
                setHtmlText(viewHolder.title, data.getTitle());
        }
        else if (holder instanceof ViewHolderItemIcon) {
            ViewHolderItemIcon viewHolder = (ViewHolderItemIcon) holder;
            PaperboyItem data = (PaperboyItem) m_dataset.get(position).data;
            ItemType definition = m_config.getItemTypes().get(data.getType());

            if (viewHolder.type == null)
                Log.w(TAG, "View id 'R.id.type' missing in custom layout");
            else {
                if (definition == null)
                    viewHolder.type.setVisibility(View.GONE);
                else {
                    viewHolder.type.setVisibility(View.VISIBLE);
                    viewHolder.type.setImageResource(definition.getIcon());
                    viewHolder.type.setColorFilter(getColor(definition));
                }
            }

            if (viewHolder.title == null)
                Log.w(TAG, "View id 'R.id.title' missing in custom layout");
            else
                setHtmlText(viewHolder.title, data.getTitle());
        }
    }

    @Override
    public int getItemViewType(int position) {
        return m_dataset.get(position).type;
    }

    @Override
    public int getItemCount() {
        return m_dataset.size();
    }

    @ColorInt
    private int getColor(@Nullable ItemType definition) {
        int color = definition == null ? 0 : definition.getColor();

        return color == 0 ? DEFAULT_COLOR : color;
    }

    @ElementType
    private int getElementType(@ViewType int type) {
        switch (type) {
            case ViewTypes.ICON:
                return ElementTypes.ITEM_ICON;
            case ViewTypes.LABEL:
                return ElementTypes.ITEM_LABEL;
            case ViewTypes.NONE:
            default:
                return ElementTypes.ITEM_NONE;
        }
    }

    private View inflateItemView(@LayoutRes int layoutRes, @NonNull ViewGroup parent) {
        if (m_config.getItemLayout() != 0)
            layoutRes = m_config.getItemLayout();

        return m_inflater.inflate(layoutRes, parent, false);
    }

    private void setHtmlText(@NonNull TextView textView, @NonNull String html) {
        Spanned spanned = Html.fromHtml(html);
        ClickableSpan[] spans = null;

        textView.setText(spanned);

        if (spanned != null)
            spans = spanned.getSpans(0, spanned.length(), ClickableSpan.class);
        if (spans != null && spans.length > 0) {
            textView.setMovementMethod(LinkMovementMethod.getInstance());

            if (!(textView instanceof TouchDispatchingTextView)) {
                Log.w(TAG, "Use " + TouchDispatchingTextView.class.getName() +
                        " for View id 'R.id.title' in your custom layout when working with HTML links");
            }
        }
        else {
            textView.setMovementMethod(null);
        }
    }
}

File: paperboy/src/main/kotlin/com/github/porokoro/paperboy/PaperboyAdapter.kt
/*
 * Copyright (C) 2015-2016 porokoro
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.porokoro.paperboy

import android.content.Context
import android.graphics.Color
import android.graphics.PorterDuff
import android.support.v7.widget.RecyclerView
import android.text.Html
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import java.util.*

internal class PaperboyAdapter(context: Context, private val config: PaperboyConfiguration)
: RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    companion object {
        val DEFAULT_COLOR = Color.BLACK
        val TAG = PaperboyAdapter::class.java.simpleName
    }

    private val inflater = LayoutInflater.from(context)
    private val dataset = arrayListOf<Element>()
    private val elementType = getElementType(config.viewType)

    fun setData(sections: List<PaperboySection>) {
        for (section in sections) {
            if (config.sortItems || config.viewType == ViewTypes.HEADER) {
                Collections.sort(section.items) { lhs, rhs ->
                    val lhsDef = config.itemTypes.get(lhs.type)
                    val rhsDef = config.itemTypes.get(rhs.type)
                    (lhsDef?.sortOrder ?: Int.MAX_VALUE) - (rhsDef?.sortOrder ?: Int.MAX_VALUE)
                }
            }

            dataset.add(Element(ElementTypes.SECTION_HEADER, section))

            var prevItemType = DefaultItemTypes.NONE

            for (item in section.items) {
                if (item.type != prevItemType && config.viewType == ViewTypes.HEADER) {
                    dataset.add(Element(
                            ElementTypes.TYPE_HEADER,
                            config.itemTypes.get(item.type)
                    ))
                    prevItemType = item.type
                }

                dataset.add(Element(elementType, item))
            }
        }

        notifyDataSetChanged()
    }

    private class ViewHolderSection(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val name = itemView.findViewById(R.id.name) as TextView?
    }

    private class ViewHolderType(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val name = itemView.findViewById(R.id.name) as TextView?
    }

    private class ViewHolderItemNone(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val title = itemView.findViewById(R.id.title) as TextView?
    }

    private class ViewHolderItemLabel(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val type = itemView.findViewById(R.id.type) as TextView?
        val title = itemView.findViewById(R.id.title) as TextView?
    }

    private class ViewHolderItemIcon(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val type = itemView.findViewById(R.id.type) as ImageView?
        val title = itemView.findViewById(R.id.title) as TextView?
    }

    private data class Element(
            val type: Int,
            val data: Any?)

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): RecyclerView.ViewHolder? = when (viewType) {
        ElementTypes.SECTION_HEADER -> {
            val layout = if (config.sectionLayout != 0) config.sectionLayout else R.layout.list_item_paperboy_section
            ViewHolderSection(inflater.inflate(layout, parent, false))
        }
        ElementTypes.TYPE_HEADER -> {
            val layout = if (config.typeLayout != 0) config.typeLayout else R.layout.list_item_paperboy_type
            ViewHolderType(inflater.inflate(layout, parent, false))
        }
        ElementTypes.ITEM_NONE -> {
            val layout = if (config.itemLayout != 0) config.itemLayout else R.layout.list_item_paperboy_item_none
            ViewHolderItemNone(inflater.inflate(layout, parent, false))
        }
        ElementTypes.ITEM_LABEL -> {
            val layout = if (config.itemLayout != 0) config.itemLayout else R.layout.list_item_paperboy_item_label
            ViewHolderItemLabel(inflater.inflate(layout, parent, false))
        }
        ElementTypes.ITEM_ICON -> {
            val layout = if (config.itemLayout != 0) config.itemLayout else R.layout.list_item_paperboy_item_icon
            ViewHolderItemIcon(inflater.inflate(layout, parent, false))
        }
        else -> null
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder?, position: Int): Unit {
        when (holder) {
            is ViewHolderSection -> {
                val data = dataset[position].data as PaperboySection

                if (holder.name == null)
                    Log.w(TAG, "View id 'R.id.name' missing in custom layout")
                else
                    holder.name.text = data.name
            }
            is ViewHolderType -> {
                val data = dataset[position].data as ItemType?

                if (holder.name == null)
                    Log.w(TAG, "View id 'R.id.name' missing in custom layout")
                else {
                    if (data == null)
                        holder.name.visibility = View.GONE
                    else {
                        holder.name.visibility = View.VISIBLE
                        holder.name.text = data.titlePlural
                        holder.name.setTextColor(data.color)
                    }
                }
            }
            is ViewHolderItemNone -> {
                val data = dataset[position].data as PaperboyItem

                if (holder.title == null)
                    Log.w(TAG, "View id 'R.id.title' missing in custom layout")
                else
                    holder.title.htmlText = data.title
            }
            is ViewHolderItemLabel -> {
                val data = dataset[position].data as PaperboyItem
                val definition = config.itemTypes.get(data.type)

                if (holder.type == null)
                    Log.w(TAG, "View id 'R.id.type' missing in custom layout")
                else {
                    if (definition == null)
                        holder.type.visibility = View.GONE
                    else {
                        holder.type.visibility = View.VISIBLE
                        holder.type.text = definition.titleSingular
                        holder.type.setBackgroundResource(R.drawable.paperboy_label_background)
                        holder.type.background.setColorFilter(getColor(definition), PorterDuff.Mode.SRC_ATOP)
                    }
                }

                if (holder.title == null)
                    Log.w(TAG, "View id 'R.id.title' missing in custom layout")
                else
                    holder.title.htmlText = data.title
            }
            is ViewHolderItemIcon -> {
                val data = dataset[position].data as PaperboyItem
                val definition = config.itemTypes.get(data.type)

                if (holder.type == null)
                    Log.w(TAG, "View id 'R.id.type' missing in custom layout")
                else {
                    if (definition == null)
                        holder.type.visibility = View.GONE
                    else {
                        holder.type.visibility = View.VISIBLE
                        holder.type.setImageResource(definition.icon)
                        holder.type.setColorFilter(getColor(definition))
                    }
                }

                if (holder.title == null)
                    Log.w(TAG, "View id 'R.id.title' missing in custom layout")
                else
                    holder.title.htmlText = data.title
            }
            else -> {
                Log.e(TAG, "Unknown view holder ${holder?.javaClass?.name}")
            }
        }
    }

    override fun getItemViewType(position: Int) = dataset[position].type

    override fun getItemCount() = dataset.size

    private fun getColor(definition: ItemType?): Int {
        val color = definition?.color ?: 0
        return if (color == 0) DEFAULT_COLOR else color
    }

    private fun getElementType(type: Int) = when (type) {
        ViewTypes.ICON -> ElementTypes.ITEM_ICON
        ViewTypes.LABEL -> ElementTypes.ITEM_LABEL
        ViewTypes.NONE -> ElementTypes.ITEM_NONE
        else -> ElementTypes.ITEM_NONE
    }

    private var TextView.htmlText: CharSequence
        get() = text
        set(value) {
            val spanned = Html.fromHtml(value.toString())
            val spans = spanned?.getSpans(0, spanned.length, ClickableSpan::class.java)

            text = spanned

            if (spans != null && spans.size > 0) {
                movementMethod = LinkMovementMethod.getInstance()

                if (this !is TouchDispatchingTextView)
                    Log.w(TAG, "Use ${TouchDispatchingTextView::class.java.name} for View id 'R.id.title' " +
                            "in your custom layout when working with HTML links")
            } else
                movementMethod = null
        }
}
